Assignment 5 - unknown language

ElizabethSvit opened the pull request.

Commit:
Assignment 5 - unknown language

ssancho2 commented on ass5.cpp:
```diff
@@ -0,0 +1,117 @@
+        adj_list[v].push_back(w);
+    }
+    
+    void topologicalSortUtil(int v, vector<bool> &visited, stack<int> &stack);
```
This one should be private

ssancho2 commented on ass5.cpp:
```diff
@@ -0,0 +1,117 @@
+        
+        for (int j = 0; j < min(str_first.length(), str_sec.length()); ++j) {
+            if (str_first[j] != str_sec[j]) {
+                graph.addEdge(str_first[j] - 'A', str_sec[j] - 'A');
```
Using this mapping from char to ints is brittle. Does it work with lower case letters, for example?

ElizabethSvit replied:
Firstly I was using " - 'a' ", but " - 'A' " it works with both case letters... Also, ASCII is easier if there is a need in unique identification of each letter. 
Better just store it as a char?

agaga23 replied:
I think it doesn't work for lower case letters. Look at 'z' in ASCII: 'z' - 'A' = 122 - 65 = 57 and  you have ALPHABET_SIZE = 50, so your adj_list.size() is too small for getting adj_list['z'-'A']. So there is a problem with your third testcase. You can check such things with valgrind if you use linux. Just write valgrind ./codename

ssancho2 replied:
Also, char can be signed or unsigned, depending on the compiler/platform. This may generate a hard to debug bug, that works on your machine, but not on production.

ssancho2 commented on ass5.cpp:
```diff
@@ -0,0 +1,117 @@
+    }
+    
+    stack<int> ordered_alphabet = graph.topologicalSort(letter_ids);
+    string output_str;
```
I think `vector<char>` or `char[]` are a bit better options for the output. If you prefer to use string, I'll suggest removing the extra spaces and just returning the alphabet chars.

ssancho2 commented on ass5.cpp:
```diff
@@ -0,0 +1,117 @@
+    }
+}
+
+void test_func() {
```
Unit testing! Nice :)

ElizabethSvit replied:
Thanks!

ssancho2 commented on ass5.cpp:
```diff
@@ -0,0 +1,117 @@
+    string output_str = return_alphabet(words);
+    string exp_output = "m a c b ";
+    
+    assert_output(output_str, exp_output, 1);
```
A comment here regarding testing this function. In this case, "m a b c " is also a valid option, right? Please see the thread on slack, but basically the idea is that if you or somebody else modifies the implementation and it produces the other (still correct) result this test will fail.

ElizabethSvit replied:
Thank you! I took into consideration the comment above and used a vector<char> for an output fnction so that the set of chars is the same and correct.

agaga23 replied:
I think that @ssancho2 meant something else. Look, for your first test case there are two correct solutions: m a c b and m a b c. We don't have enough knowledge to decide which one is real. So your assert_output should write OK for both of them. So you could give all solutions as exp_output (vector<vector<char> >) or check if the solution which your program found keeps all  dependencies from the given dictionary.

ssancho2 replied:
Yes, exactly.
In practice, we do this a lot, specially when otherwise the test will be as complex as the code tested :), but it's good to know that it's not ideal and the technical debt we acquire.

ssancho2 reviewed (commented):
Some comments

ssancho2 commented on ass5.cpp:
```diff
@@ -0,0 +1,117 @@
+    return stack;
+}
+
+string return_alphabet(vector<string> words) {
```
What happens if the input is `{"ABC"}` ?

ElizabethSvit replied:
So I guess if there is only one word and nothing else to compare with lexicographically, any letter order is possible. For now I am comparing it only with one variant, but probably it's better to make some assertion with all cases together.

Commit:
Updated ass5

Added a test case with one word and changed output of a function to a vector of chars to make assertion universal.

agaga23 commented:
You have a different approach than me :) It was a little bit challenging for me to understand why it works, but now I agree, it is ok :) Only one additional comment.

ssancho2 commented:
Note that if the code is hard to understand it's ok to ask for simplification and/or more comments. The purpose of the reviews it to ensure that the code is correct AND easy to be taken over by somebody else and keep working on it.

Commit:
Assignment 6 - Rearranging cars

ElizabethSvit commented:
Thank you for your comments! Somehow the solution of assignment 6 appeared here accidentially. I'm merging this pull request so that I can start a new one with assignment 6.
